pallets/order-book/src/lib.rs

```diff
@@ -0,0 +1,639 @@
+		/// Provided for frontend to grab orders for an individual account
+		pub fn get_account_orders(
+			account_id: T::AccountId,
+		) -> Result<
+			sp_std::vec::Vec<(
+				T::Hash,
+				Order<T::Hash, T::AccountId, T::AssetCurrencyId, T::ForeignCurrencyBalance>,
+			)>,
+			Error<T>,
+		> {
+			Ok(<UserOrders<T>>::iter_prefix(account_id).collect())
+		}
+
+		/// Get all orders
+		/// Provided for frontend to grab all open orders
+		pub fn get_all_orders() -> Result<
+			sp_std::vec::Vec<(
+				T::Hash,
+				Order<T::Hash, T::AccountId, T::AssetCurrencyId, T::ForeignCurrencyBalance>,
+			)>,
+			Error<T>,
+		> {
+			Ok(<Orders<T>>::iter().collect())
+		}
+
+		/// Remove an order from storage
+		pub fn remove_order(order_id: T::Hash) -> DispatchResult {
+			let order = <Orders<T>>::get(order_id)?;
+			<UserOrders<T>>::remove(&order.placing_account, order.order_id);
+			<Orders<T>>::remove(order.order_id);
+			let mut orders = <AssetPairOrders<T>>::get(order.asset_in_id, order.asset_out_id);
+			orders.retain(|o| *o != order.order_id);
+			<AssetPairOrders<T>>::insert(order.asset_in_id, order.asset_out_id, orders);
+			Ok(())
+		}
+
+		/// Generate a hash to be used for the order id using the placing
+		/// account, asset IDs, and pallet nonce.
+		pub fn gen_hash(
+			placer: &T::AccountId,
+			asset_out: T::AssetCurrencyId,
+			asset_in: T::AssetCurrencyId,
+			nonce: T::Nonce,
+		) -> T::Hash {
+			(&placer, asset_in, asset_out, nonce).using_encoded(T::Hashing::hash)
+		}
+	}
+
+	impl<T: Config> TokenSwaps<T::AccountId> for Pallet<T>
+	where
+		<T as frame_system::Config>::Hash: PartialEq<<T as frame_system::Config>::Hash>,
+	{
+		type Balance = T::ForeignCurrencyBalance;
+		type CurrencyId = T::AssetCurrencyId;
+		type OrderId = T::Hash;
+
+		/// Creates an order.
+		/// Verify funds available in, and reserve for  both chains fee currency
+		/// for storage fee, and amount of outgoing currency as determined by
+		/// the buy amount and price.
+		fn place_order(
+			account: T::AccountId,
+			currency_in: T::AssetCurrencyId,
+			currency_out: T::AssetCurrencyId,
+			buy_amount: T::ForeignCurrencyBalance,
+			sell_price_limit: T::ForeignCurrencyBalance,
+			min_fullfillment_amount: T::ForeignCurrencyBalance,
+		) -> Result<Self::OrderId, DispatchError> {
+			ensure!(currency_in != currency_out, Error::<T>::ConflictingAssetIds);
+
+			ensure!(
+				buy_amount != T::ForeignCurrencyBalance::zero(),
+				Error::<T>::InvalidBuyAmount
+			);
+			ensure!(
+				sell_price_limit != T::ForeignCurrencyBalance::zero(),
+				Error::<T>::InvalidMinPrice
+			);
+			ensure!(
+				T::AssetRegistry::metadata(&currency_in).is_some(),
+				Error::<T>::InvalidAssetId
+			);
+			ensure!(
+				T::AssetRegistry::metadata(&currency_out).is_some(),
+				Error::<T>::InvalidAssetId
+			);
+			ensure!(
+				T::ReserveCurrency::can_reserve(
+					&account,
+					T::Fees::fee_value(T::OrderFeeKey::get())
+				),
+				Error::<T>::InsufficientReserveFunds,
+			);
```

Comment:
I was thinking it would be cheaper to try and check before reserve, but we can also remove if wanted as well.

Reply:
No strong opinion here 👍🏻. It's just I'm always checking lines that could be removed 😆.

To give my reasoning, I think in case of an error, the weight of this call will be the same, which is higher than the computation time the node wasted on it for an error. The weight will be added to the current block weight as if the call was ok. So, from my understanding, we would not be saving computation or making it cheaper.

Reply:
All of the reserve checks except the reserve check before the transfers to ensure enough funds can actually be transferred were removed.  Also just Found/fixed an error with fund check & added tests when double checking again for extra superfluous checks.